File: 29Items-Projects/7Algorithms/6-Backtracking/python/subset_sum.py
"""Subset Sum — find all subsets that sum to a target value."""


class SubsetSum:
    def __init__(self, numbers: list[int], target: int):
        self.numbers = numbers
        self.target = target
        self.solutions: list[list[int]] = []

    def solve_and_print(self) -> None:
        print(f"Set: {self.numbers}")
        print(f"Target sum: {self.target}\n")

        self._backtrack(0, 0, [])

        if not self.solutions:
            print(f"No subset sums to {self.target}.")
        else:
            print(f"Found {len(self.solutions)} subset(s):")
            for i, sol in enumerate(self.solutions, 1):
                print(f"  #{i}: {sol} = {self.target}")

    def _backtrack(self, index: int, current_sum: int, current: list[int]) -> None:
        if current_sum == self.target and current:
            self.solutions.append(current[:])
        for i in range(index, len(self.numbers)):
            if current_sum + self.numbers[i] <= self.target:
                current.append(self.numbers[i])
                self._backtrack(i + 1, current_sum + self.numbers[i], current)
                current.pop()

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/subsetsum/SubsetSum.java
package subsetsum;

import java.util.*;

public class SubsetSum {

    private final int[] set;
    private final int target;
    private final List<List<Integer>> solutions = new ArrayList<>();

    public SubsetSum(int[] set, int target) {
        this.set = set;
        this.target = target;
    }

    public void solveAndPrint() {
        System.out.println("Set: " + Arrays.toString(set));
        System.out.println("Target sum: " + target);
        System.out.println();

        backtrack(0, 0, new ArrayList<>());

        if (solutions.isEmpty()) {
            System.out.println("No subset sums to " + target + ".");
        } else {
            System.out.printf("Found %d subset(s):%n", solutions.size());
            for (int i = 0; i < solutions.size(); i++) {
                System.out.printf("  #%d: %s = %d%n", i + 1, solutions.get(i), target);
            }
        }
    }

    private void backtrack(int index, int currentSum, List<Integer> current) {
        if (currentSum == target && !current.isEmpty()) {
            solutions.add(new ArrayList<>(current));
        }
        for (int i = index; i < set.length; i++) {
            if (currentSum + set[i] <= target) {
                current.add(set[i]);
                backtrack(i + 1, currentSum + set[i], current);
                current.remove(current.size() - 1);
            }
        }
    }
}
